### ISSUE TITLE
autopilot is archived

### ISSUE THREAD

**Issue by FlorianSW** [nulldriver/cf-cli-resource#54]
This is just an information, as I'm not sure if and what we can do now, however, the autopilot plugin (https://github.com/contraband/autopilot) is archived now and will not get any further updates :( This plugin is used in this repo to allow a zero-downtime-deployment. Maybe we find another plugin which can do the same or something like that?

**Comment by louisthomas** [comment]
Maybe we could use: cf v3-zdt-push
https://docs.cloudfoundry.org/devguide/deploy-apps/zero-downtime.html

**Comment by patrickcrocker** [comment]
Very interesting... My main concern with this being archived would be if any future changes in the `cf` cli would require changes to the `autopilot` plugin to be compatible.  If we ever run into this situation, we'll know immediately since the `cf-cli-resource` build pipeline and/or integration tests should fail.

Regardless, I'm open to including other `cf` cli plugins into this resource.

**Comment by FlorianSW** [comment]
My main point is, that the autopilot plugin does not support the `buildpacks` property of the manifest, so I need to use the deprecated `buildpack` one. If this can be fixed in any other way, I'm open :)

**Comment by patrickcrocker** [comment]
@louisthomas I really like what they are doing with `cf v3-zdt-push` (and I've successfully manually tested it with my PCF env I use for `cf-cli-resource` integration testing) but it does not support using `manifest.yml` files yet which makes it not a great replacement for `autopilot` since it treats the manifest as the single source of truth for the app deployment.  I'm happy to investigate implementing `v3-zdt-push` in a separate thread (where we can debate the merits of implementing a beta feature).

@FlorianSW to your point, without `autopilot` being able to properly support the new `buildpacks:` manifest attribute... we'd have to hack something in the `cf-cli-resource` to translate `buildpacks` to the deprecated `buildpack`... but then what happens if someone actually did put in multiple `buildpacks`... which one would we use? (last one wins? error out if more than one buildpack is specified?)

Open to ideas here.

**Comment by FlorianSW** [comment]
@patrickcrocker I don't think that it makes sense to find out own way to handle `buildpacks` on our own (e.g. with what you said, including all of the mentioned problems). I think, it would make more sense to find a good replacement for the autopilot plugin, or even better wait until the v3-zdt-push is "finished" (finished = supporting also the manifest file). For now, the autopilot plugin works just fine, and as long as it works in some way (even by using the deprecated `buildpack` option) is something, which is ok I think :]

**Comment by aegershman** [comment]
+1 to @FlorianSW it seems the best option is to hold out to see what the landscape will be, especially for native cf-cli options like the v3-zdt-push. Autopilot is archived but it's not going away, so there's no absolute immediacy. And if it turns out v3-zdt-push isn't such a great option, then it makes more sense for a new plugin to be written (or some other orchestration tool for pushing apps?) than have the cf-cli-resource do hacky logic. 

I feel like automation is at it's best when it simply coordinates tools, like the cf-cli or plugins, rather than tries to implement too much logic on it's own (like doing something hacky with the `buildpacks` attribute)

in other words, _do nothing_, at least for the time being 😉

**Comment by patrickcrocker** [comment]
Doing nothing for the time being  😉

**Comment by eruvanos** [comment]
Hi @patrickcrocker,
I just saw, that concourse/cf-resource replaced the autopilot with an own implementation.

https://github.com/concourse/cf-resource/blob/6b9ec74dde53efa64d38c2aa1757bad74050eb8e/out/zdt/push.go

Would it be time to move on and maybe do the same?

**Comment by patrickcrocker** [comment]
@eruvanos Yes, let's discuss what this might look like.

**Comment by tenjaa** [comment]
Should we provide a pull request draft?
Then we would have something specific to discuss.

**Comment by patrickcrocker** [comment]
Thanks to a PR by @shyamz-22, we've now got a hand-crafted `zero-downtime-push` implementation.  I took that great work and completely re-tooled the integration test to ensure that we are covering all aspects of the code.  I've also added missing functionality to make it a complete replacement for the `cf-resource`.

The new code is currently available in release candidate v2.19.0-rc.2+

If you want to review the code changes, take a look at these two files:
- [itest/run-app-zero-downtime-push-tests](https://github.com/nulldriver/cf-cli-resource/blob/master/itest/run-app-zero-downtime-push-tests)
- [resource/commands/zero-downtime-push.sh](https://github.com/nulldriver/cf-cli-resource/blob/master/resource/commands/zero-downtime-push.sh)

My call to action is for people to try out the new implementation before I release this into the wild.  The easiest way to do this is to switch your pipeline to the release candidate docker repo:
```yaml
- name: cf-cli-resource
  type: registry-image
  source:
    repository: nulldriver/cf-cli-resource-rc
    tag: latest
```

Please provide your feedback.  Thank you all!

@FlorianSW @louisthomas @aegershman @eruvanos @tenjaa @shyamz-22

**Comment by shyamz-22** [comment]
@patrickcrocker Thank you for the update,  I will try the release candidate and give you feedback.

Just wanted to mention that the rewrite of cf-cli-resource makes it a breeze to add additional functionality. Best modularised bash code I have come across. Thank you for your effort  😸

**Comment by FlorianSW** [comment]
I also switched to the release candidate version in one of our pipelines, and so far it is working like a Charme.

One Thing, we don't use the zero-downtime deployment in this pipeline anymore, as we need to use a v3 command on the app during the deployment.

However, I'll switch to the RC version on another pipeline soon, where we use the zero-downtime-deployment, and give feedback then as well.

So far, from what I saw on the PR, that looks pretty much the same what we did in our current pipeline as well (except our v3 command), which is promising I think :P

**Comment by patrickcrocker** [comment]
> Just wanted to mention that the rewrite of cf-cli-resource makes it a breeze to add additional functionality. Best modularised bash code I have come across. Thank you for your effort 😸

Thank you for the feedback, I really appreciate that!

**Comment by tenjaa** [comment]
Works also great for us. We even tested (involuntarily) the rollback on failed deployments which also works like a charm!

**Comment by FlorianSW** [comment]
So far, we did not experience any issues so far :)

**Comment by patrickcrocker** [comment]
Implemented in [v2.19.0](https://github.com/nulldriver/cf-cli-resource/releases/tag/v2.19.0)!
